tensorboard/webapp/routes/core_deeplink_provider_test.ts

```diff
@@ -61,175 +66,222 @@ describe('core deeplink provider', () => {
   });
 
   describe('time series', () => {
-    it('serializes pinned card state when store updates', () => {
-      store.overrideSelector(selectors.getPinnedCardsWithMetadata, [
-        {
-          cardId: 'card1',
-          plugin: PluginType.SCALARS,
-          tag: 'accuracy',
-          runId: null,
-        },
-      ]);
-      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, [
-        {
-          plugin: PluginType.SCALARS,
-          tag: 'loss',
-        },
-      ]);
-      store.refreshState();
+    describe('smoothing state', () => {
+      it('serializes the smoothing state to the URL', () => {
+        store.overrideSelector(selectors.getMetricsScalarSmoothing, 0);
+        store.refreshState();
 
-      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual([
-        {
-          key: 'pinnedCards',
-          value:
-            '[{"plugin":"scalars","tag":"accuracy"},{"plugin":"scalars","tag":"loss"}]',
-        },
-      ]);
-
-      store.overrideSelector(selectors.getPinnedCardsWithMetadata, [
-        {
-          cardId: 'card1',
-          plugin: PluginType.SCALARS,
-          tag: 'accuracy2',
-          runId: null,
-        },
-      ]);
-      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, [
-        {
-          plugin: PluginType.SCALARS,
-          tag: 'loss2',
-        },
-      ]);
-      store.refreshState();
+        expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual(
+          [
+            {
+              key: 'smoothing',
+              value: '0',
+            },
+          ]
+        );
+      });
 
-      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual([
-        {
-          key: 'pinnedCards',
-          value:
-            '[{"plugin":"scalars","tag":"accuracy2"},{"plugin":"scalars","tag":"loss2"}]',
-        },
-      ]);
-    });
+      it('does not reflect default value to the URL', () => {
+        store.overrideSelector(selectors.getMetricsScalarSmoothing, 0.6);
+        store.refreshState();
 
-    it('serializes nothing when states are empty', () => {
-      store.overrideSelector(selectors.getPinnedCardsWithMetadata, []);
-      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, []);
-      store.refreshState();
+        expect(queryParamsSerialized.length).toBe(0);
+      });
 
-      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual(
-        []
-      );
-    });
+      it('deserializes the state in the URL without much sanitization', () => {
```

Comment:
I'm not sure if we have any need to pass the raw, unsanitized smoothing value from the URL all the way to the reducer.  Would it be better to do the isFinite check, etc here in the core_deeplink_provider, given that `pinnedCards` is also sanitized here?

Reply:
I am inclined to keep it as is. That is because at the end of the day, reducer should do more input validations anyways since we really don't to spread validation logic all over the places and reducer is where the business logic should normally go.

Even without `Number.isFinite` call in the reducer and in the provider, since it is `null`able, we still need to do similar kind of guards in the reducer.

Reply:
That's fair enough.  If this is our stance, perhaps some of the `pinnedCards` sanitization logic should really belong in its reducer as well, then (unrelated to this PR of course).

Reply:
In fact, I would have added the smoothing change validation logic in the reducer, too :)